samples: tests task wdt on some stm32 nucleo target boards

FRASTM opened the pull request:
The IWDG timeout depends on the actual LSI clock frequency which feeds the watchdog timer.
On the stm32l152, stm32f091, stm32l073, stmf429, due to manufacturing the LSI clock is given with a typical value and a min-Max range. The IWDT is clocked by the LSI which has a frequency of 40kHz for those stm32 devices. The LSI freq must be adjusted to its real value to give the IWDG timeout.
We add an overlay for this sample on the target board to set a more accurate LSI freq. higher than the default 37kHz and this should be updated for any board
once the actual LSI freq differs from the  LSI default value set by the clk_lsi node of the DTS:
```
 		clk_lsi: clk-lsi {
			#clock-cells = <0>;
			compatible = "fixed-clock";
			clock-frequency = <DT_FREQ_K(37)>;
			status = "disabled";
		};

```
Other changes are applied coming from the stm32Cube example.
The driver should enable the IWDG just before changing the prescaler and reload registers, else it starts running immediately.
After updating the IDT prescaler and reload registers, the IWDT SR gives the position of the PVU and the RVU bits thanks to a LL function. LL_IWDG_IsReady() returns 1 when the update is completed, so the driver should wait until LL_IWDG_IsReady is not 0 anymore. There is no need to include a timeout.
The RefMan of stm32l073, stm32l152, stm32f091, stm32f429 is confusing because it explicitly mentions "5 RC 40kHz cycles" 
The correct formula is **t_IWDG(ms) = t_LSI(ms) x 4 x 2^(IWDG_PR[2:0]) x (IWDG_RLR[11:0] + 1**)
There is no need to enable the LSI explicitely in the iwdt driver.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/43024

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Note : according to the ErrataSheet of the STM32L15xxE and STM32L15xVD  [ES0235](https://www.st.com/resource/en/errata_sheet/es0235-stm32l15xxe-and-stm32l15xvdx-ultralowpower-device-limitations-stmicroelectronics.pdf)
The RVU and PVU flags of the IWDG Satus Reg SR  are not reset in Stop mode

FRASTM committed:
drivers: watchdog stm32 iwdg has a long update register delay

Check the update of the prescaler and reload register in the status Reg
before setting a new timeout. This is done with the LL_IWDG_IsReady()
function that returns 1 when the update is completed
else a timeout occurs. Enabling the IWDG comes (with setup function)
after writing the prescaler and counter reg.
Note that in stm32L0x, stm32L1x, stm32F0x, stm32F4x the LSI clock is 37kHz.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: stm32 independent watchdog enable lsi clock

Adding the freq value of the LSI clock source STM32_MAX_LSI_FREQ
when calculating the timeout of the iwdt independant watchdog
That value is given by the DTS or could be the typical LSI freq.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
include: drivers clock control stm32 lsi freq value

Define the STM32_LSI_FREQ as the LSI freq value
of the LSI clock from the DTS. with IWDG, no need to enable LSI.
According to the stm32x DataSheet this value depends on the soc
and can vary in a wide range. LSI1 for stm32wbx.
LSI clock freq. should not come from the stm32Cube driver.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
twister -T samples/subsys/task_wdt  on all available target boards :
```
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f746zg,passed,,na,2.2868571281433105,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_wb55rg,passed,,na,2.438084602355957,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_g071rb,passed,,na,2.2301554679870605,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,stm32l562e_dk,passed,,na,3.400723934173584,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_g474re,passed,,na,2.3637142181396484,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_wl55jc,skipped,,na,,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,stm32f3_disco,passed,,na,3.5216867923736572,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f429zi,passed,,na,2.7286908626556396,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_h743zi,passed,,na,2.8374483585357666,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_l152re,passed,,na,10.565778017044067,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,b_u585i_iot02a,passed,,na,4.118236064910889,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f091rc,passed,,na,3.2924914360046387,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f103rb,passed,,na,3.1735715866088867,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,disco_l475_iot1,passed,,na,7.015970945358276,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_l073rz,passed,,na,7.33816933631897,0,0
```

FRASTM commented:
enable the sample and iwdg test on the nucleo_wl55jc board:
`samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_wl55jc,passed,,na,7.319908380508423,0,0`

ABOSTM commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -24,14 +25,18 @@
 #define IWDG_RELOAD_MIN		(0x0000U)
 #define IWDG_RELOAD_MAX		(0x0FFFU)
 
+#ifndef STM32_LSI_CLOCK
+#define STM32_LSI_CLOCK LSI_VALUE
+#endif /* STM32_LSI_CLOCK */
+
```
This default value correspond to #else in stm32_clock_control.h

erwango replied:
Right, let's centralize symbol definitions in one single file

erwango commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -119,6 +119,10 @@
 
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi), fixed_clock, okay)
+#define STM32_LSI_CLOCK		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
+#endif
```
I prefer we avoid mixing Cube and Zephyr defines

ABOSTM replied:
but with a risk to have 2 different values STM32_LSI_CLOCK != LSI_VALUE  for the same thing

erwango replied:
Thinking to it a bit more, let's add a #else case for the time being, but please add a #warning message to inform that LSI clock_freq should be defined in device tree as a deprecation notice.

FRASTM commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -188,7 +191,7 @@ static int iwdg_stm32_init(const struct device *dev)
 	 * the LSI oscillator is forced ON and cannot be disabled.
 	 *
-	 * t_IWDG(ms) = t_LSI(ms) x 4 x 2^(IWDG_PR[2:0]) x (IWDG_RLR[11:0] + 1)
+	 * t_IWDG(ms) = t_LSI(ms) x 4096 x 2^(IWDG_PR[2:0]) x (IWDG_RLR[11:0] + 1)
```
that's right, keeping initial formula

FRASTM commented:
keeping formula : t_IWDG(ms) = t_LSI(ms) x 4 x 2^(IWDG_PR[2:0]) x (IWDG_RLR[11:0] + 1)

FRASTM commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -137,18 +136,14 @@ static int iwdg_stm32_install_timeout(const struct device *dev,
-			return -ENODEV;
-		}
+	/* Wait for the update operation completed : expecting IWDG to be ready */
+	while (LL_IWDG_IsReady(iwdg) != 1) {
```
Ok, keeping the timeout is possible, but not really useful. The prescaler and counter registers _will_ be modified. After the update, : RVU and PVU update bits of the IWDG status reg bits will retrieve 0, and consequently LL_IWDG_IsReady will become true, no doubt.
 It is just a question of waiting enough (IWDG_SR_UPDATE_TIMEOUT). A timeout is required if bits could remain (be stuck) 1. 

BTW why returning -ENODEV instead of -ETIMEDOUT  ?

FRASTM commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -104,7 +104,10 @@ static int iwdg_stm32_setup(const struct device *dev, uint8_t options)
 	}
 
-	LL_IWDG_Enable(iwdg);
+	/*
+	 * do not enable iwdg yet else it cannot be stopped
+	 * and counter has default timeout
+	 */
```
Correct, however the stm32 IWDG behaviour (at least with stm32l152) is that modifying the Prescaler and Reload counter requires the LL_IWDG_Enable before the LL_IWDG_EnableWriteAccess or the LL_IWDG_IsReady (RVU and PVU status register bits remains high and becomes low once the IWDG is enabled.
In other words, the IWDG update is ongoing until IWDG is enabled  : Prescaler and Reload can be modified but status (PVU, RVU) bits requires IWDG enable to be checked. This is the purpose of the Note in the ref Man. 
"_If several reload values or prescaler values are used by application, it is mandatory to wait
until RVU bit is reset before changing the reload value and to wait until PVU bit is reset
before changing the prescaler value. However, after updating the prescaler and/or the
reload value it is not necessary to wait until RVU or PVU is reset before continuing code
execution_"
--> checking RVU and PVU status bit is mandatory before writing the Prescaler and Reload, not after.

As long as the `iwdg_stm32_install_timeout`  (writing RLR and PR ) should come **before** the `iwdg_stm32_setup` (enabling IWDG), the check of RVU and PVU bits should come in the iwdg_stm32_setup function, after enabling the IWDG. 

LL_IWDG_Enable removed from _init functions only

FRASTM replied:
Iin the `iwdg_stm32_init`, only  if  NOT CONFIG_WDT_DISABLE_AT_BOOT we should have `iwdg_stm32_install_timeout` followed by `iwdg_stm32_setup`
else there is no reason that iwdg_stm32_init configures and runs the IWDG.

FRASTM commented:
change to control the iwdg_stm32_install_timeout with prescaler and counter register update outside the  iwdg_stm32_setup
which just starts the IWDG.

FRASTM commented:
rebase
+ define the STM32_LSI_CLOCKin the stm32_clock_control.h

erwango commented:
> rebase define the STM32_LSI_CLOCK in the stm32_clock_control.h add a comment to warn about LSI clock definition --> #warning **treated as error** cannot pass zephyr tests

`#warning ` case should be faced only by users that didn't defined LSI clock freq by device tree.
In tree, all cases occurrences should be converted to device tree.

FRASTM commented:
> > rebase define the STM32_LSI_CLOCK in the stm32_clock_control.h add a comment to warn about LSI clock definition --> #warning **treated as error** cannot pass zephyr tests
> 
> `#warning ` case should be faced only by users that didn't defined LSI clock freq by device tree. In tree, all cases occurrences should be converted to device tree.

In case of IWDG node enable, the LSI is automatically enabled, getting the the LSI_freq.
The warning comes when the LSI_CLOCK is read from HAL stm32Cube driver when the lsi_clk node is not relevant (no IWDG and no clk_lsi enabled)

FRASTM committed:
boards: arm: nucleo_wl55jc includes the iwdg node

Enables the independent watchdog on the nuclo_wl55 target board

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on adc901aa6a39caffd7971ff99456b26414cd3793
raise a warning if LSI clock freq is not defined by the DTS

FRASTM commented:
keep STM32_LSI_FREQ definition for the LSI freq (instead of STM32_LSI_CLOCK)

FRASTM commented:
test results : 
```
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f746zg,passed,,na,5.80250096321106,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_wb55rg,passed,,na,5.7885401248931885,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_g071rb,passed,,na,5.699979543685913,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,stm32l562e_dk,passed,,na,6.760100841522217,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_g474re,passed,,na,6.065260887145996,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_wl55jc,passed,,na,7.526520013809204,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,stm32f3_disco,passed,,na,7.266631126403809,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f429zi,passed,,na,6.1810266971588135,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_h743zi,passed,,na,6.286457777023315,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_l152re,passed,,na,10.450675964355469,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,b_u585i_iot02a,passed,,na,8.260071992874146,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f091rc,passed,,na,8.25836968421936,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_f103rb,passed,,na,8.3711256980896,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,disco_l475_iot1,passed,,na,6.899222135543823,0,0
samples/subsys/task_wdt/sample.subsys.task_wdt,arm,nucleo_l073rz,passed,,na,11.71114706993103,0,0
```

erwango commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -164,12 +164,29 @@
 #endif
 
-#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi), fixed_clock, okay)
+#if defined(CONFIG_SOC_SERIES_STM32WBX)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_watchdog)
+/* IWDG automatically enables the LSI clock : get its freq. */
+#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
+#elif DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi1), fixed_clock, okay)
+#define STM32_LSI_ENABLED	1
+#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
+#else
+/* LSI clock freq should be in device tree, not from stm32Cube LSI_VALUE */
+#define STM32_LSI_FREQ		0
+#endif
+#else /* CONFIG_SOC_SERIES_STM32WBX */
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_watchdog)
+/* IWDG automatically enables the LSI clock : get its freq. */
+#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
+#elif DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi), fixed_clock, okay)
 #define STM32_LSI_ENABLED	1
 #define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
 #else
+/* LSI clock freq should be in device tree, not from stm32Cube LSI_VALUE */
 #define STM32_LSI_FREQ		0
 #endif
+#endif /* CONFIG_SOC_SERIES_STM32WBX */
```
Please review this section.
1- LSI is common to the whole system, so is this file. We should not see driver specific code in this file (`st_stm32_watchdog` in this case).
2- If series specific code is required, please use compatilble to catch with what we've done till now in this file (no `CONFIG_SOC_SERIES_STM32WBX`)

FRASTM replied:
1. The iwdt is a special case where the frequency of the LSI (or LSI1) RC oscillator is needed from the **clk-lsi node** of the soc DTS even if this node is not enabled in the board.
The STM32_LSI_FREQ is exactly what is needed by the iwdg to calculate its timeout value. However we could still define a STM32_LSI_CLOCK for the iwdg purpose only.

2. what about devices like stm32wb with LSI 1 (and LSI2) RC oscillator instead of LSI : the STM32_LSI_FREQ is the LSI1 RC oscillator. Should the STM32_LSI_FREQ be defined as well

FRASTM replied:
using the LSI clock for iwdt purpose could be :
```
/* the STM32_LSI_FREQ is given by the LSI or LSI1 RC oscillator */
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_watchdog)
/* when IWDG is on, the LSI is not necessarily enabled, just get its freq. */
#if DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
#elif DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
#else
/* LSI clock freq should be in device tree, not from stm32Cube LSI_VALUE */
#define STM32_LSI_FREQ		0
#endif
#else /* st_stm32_watchdog */
 #if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi), fixed_clock, okay)
 #define STM32_LSI_ENABLED	1
 #define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
#elif DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_lsi1), fixed_clock, okay)
#define STM32_LSI_ENABLED	1
#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
 #else
 #define STM32_LSI_FREQ		0
 #endif
#endif /* st_stm32_watchdog */
 
 #if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_hsi), st_stm32h7_hsi_clock, okay)
 #define STM32_HSI_ENABLED	1
```

erwango commented on samples/subsys/task_wdt/boards/nucleo_f091rc.overlay:
```diff
@@ -0,0 +1,9 @@
+ * Copyright (c) 2022 STMicroelectronics
+ */
+
+&clk_lsi {
+	clock-frequency = <DT_FREQ_K(40)>;
+};
```
LSI clock freq is not wdt specific.
Should it be put in board.dts ? If not a comment is required to explain this choice

FRASTM replied:
with comment
```
 /* 
 * Due to the LSI RC oscillator characteristics (defined by the datasheet),
 * the typical value set in the soc or board dts might be too low
 * for the IWWDG timeout : define the max value for this usecase.
 */
```

FRASTM commented:
- remove reference to any driver in the  include/drivers/clock_control/stm32_clock_control.h 
- get the LSI frequency from LSI or LSI1 RC oscillator
- define LSI enable depending on the clk_lsi (or clk_lsi1) enable

martinjaeger commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -164,12 +164,20 @@
+#define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
 #else
+/* LSI clock freq should be in device tree, not from stm32Cube LSI_VALUE */
 #define STM32_LSI_FREQ		0
```
Frequency of `0` is not a safe default and will now lead to a division by zero (hard fault) in `drivers/watchdog/wdt_iwdg_stm32.c`.

FRASTM replied:
In this case the freq could be the stm32 HAL one 
```
#warning "LSI clock freq should be in device tree, not from stm32Cube"
#define STM32_LSI_FREQ		LSI_VALUE
```

martinjaeger commented on samples/subsys/task_wdt/boards/nucleo_f091rc.overlay:
```diff
@@ -0,0 +1,14 @@
+ * Copyright (c) 2022 STMicroelectronics
+ */
+
+/*
+ * Due to the LSI RC oscillator characteristics (defined by the datasheet),
+ * the typical value set in the soc or board dts might be too low
+ * for the IWDG timeout : define the max value for this usecase.
+ */
+&clk_lsi {
+	clock-frequency = <DT_FREQ_K(50)>;
+};
```
If this overlay is required always if the task watchdog is used, this is a very hacky approach and not a valid solution. The IWDG and task watchdog must work with default IWDG settings coming from SOC .dts. Either the task watchdog has to be changed to accept higher tolerances or this setting must be configured for *all* relevant STM32 MCUs in the SOC .dts files, not in an overlay. And it should probably called `max-clock-frequency` in that case, not `clock-frequency`, which is the nominal/average if I understand correctly.

FRASTM replied:
Only the conversion of IWDG timeout in ticks, in iwdg_stm32_convert_timeout function, requires the LSI MAXimum value as given by the datasheet (typical LSI freq might be too low in for several devices). 

```
	/* Convert timeout to LSI clock ticks. */
	uint32_t ticks = (uint64_t)timeout * STM32_LSI_FREQ / USEC_PER_SEC;
```

In the <clk-lsi> node of the soc DTS, we could add a <MAX-clock-frequency> property so that IWDG driver use that value in the calculation as above --> impacting the stm32 IWDG driver only
Or we always give the MAX value to the <clock-frequency> of the <clk-lsi> node --> impacting all the users of lsi freq.

FRASTM replied:
Introducing a compatible: "st,stm32-lsi-clock" for all the stm32 devices with a `<max-clock-frequency>` 

then will the wdt_idwg_stm32.c use the STM32_MAX_LSI_FREQ
```
/* Convert timeout to LSI clock ticks. */
uint32_t ticks = (uint64_t)timeout * STM32_LSI_FREQ / USEC_PER_SEC;
```

FRASTM committed:
dts: bindings: introduce a stm32 lsi-clock with an optional max freq

Some stm32 mcus have a wide range of LSI oscillator freq, where the
typical value can be very different from the Max value (given by the
device Datasheet).
Some drivers require this Max freq instead of the typical one.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: stm32: soc definition of lsi clock is compatible st,stm32-lsi-clock

Change the DTS node of the clk_lsi for each stm32 device to be
compatible with the st,stm32-lsi-clock.
It just adds the max-clock-frequency property as optional for
some stm32 mcus where the RC oscillator characteristics requires it.
For the stm32wb, map another lsi node to lsi1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -164,11 +164,19 @@
+/* the STM32_LSI_FREQ is given by the LSI or LSI1 RC oscillator, even if not enabled */
+#if DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
 #define STM32_LSI_FREQ		DT_PROP(DT_NODELABEL(clk_lsi), clock_frequency)
+#elif DT_PROP(DT_NODELABEL(clk_lsi1), clock_frequency)
```
Instead of adding LSI1, we can tweak wb.dtsi to set 'lsi' as an alias to `lsi1` in the following way:
```
      lsi1: lsi@address {
          bla;
          foo;
      };
```

FRASTM replied:
is it 
```
		clk_lsi: clk-lsi {
			#clock-cells = <0>;
			compatible = "st,stm32-lsi-clock";
			clock-frequency = <DT_FREQ_K(32)>;
			status = "disabled";
		};
```
in the dts/arm/st/wb/stm32wb.dtsi

FRASTM replied:
defining the lsi1 of the stm32wb as lsi node in its DTS:
```
clk_lsi: clk-lsi1 {
...
}
```
So there is only one reference  **DT_NODELABEL(clk_lsi)**   in the driver

FRASTM commented:
rebase on 736520a4fef642e4cfb8ee2e273486f73e0ea2aa

martinjaeger commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -60,7 +69,7 @@ static void iwdg_stm32_convert_timeout(uint32_t timeout,
 
 	/* Convert timeout to LSI clock ticks. */
-	uint32_t ticks = (uint64_t)timeout * LSI_VALUE / USEC_PER_SEC;
+	uint32_t ticks = (uint64_t)timeout * STM32_MAX_LSI_FREQ / USEC_PER_SEC;
```
I'm still wondering a bit if this is what the user would expect. If a user reads the datasheet and already considers the tolerance in the timeout selection, he/she would get a significantly longer timeout than expected (on average).

Did you consider adjusting below Kconfig instead to make the task watchdog pass tests for the mentioned board?

https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/task_wdt/Kconfig#L47-L60

FRASTM replied:
absolutely, adjusting the  CONFIG_TASK_WDT_HW_FALLBACK_DELAY=100  for the boards can also fix that

FRASTM replied:
The TASK_WDT_MIN_TIMEOUT is 1..10000, with a LSI clock variation from 37000 (typical) to 56000 (max), the CONFIG_TASK_WDT_HW_FALLBACK_DELAY cab be given by :

-    _upper value of TASK_WDT_MIN_TIMEOUT * LSI_CLOCK typical freq / USEC_PER_SEC - upper value of TASK_WDT_MIN_TIMEOUT * LSI_CLOCK max freq / USEC_PER_SEC_

depending on the soc: 
stm32F0x : CONFIG_TASK_WDT_HW_FALLBACK_DELAY=100
stm32F4x : CONFIG_TASK_WDT_HW_FALLBACK_DELAY=200
stm32L1x : CONFIG_TASK_WDT_HW_FALLBACK_DELAY=200
stm32L0x : CONFIG_TASK_WDT_HW_FALLBACK_DELAY=200

FRASTM commented on dts/arm/st/f1/stm32f1.dtsi:
```diff
@@ -55,7 +55,7 @@
 			#clock-cells = <0>;
-			compatible = "fixed-clock";
+			compatible = "st,stm32-lsi-clock";
 			clock-frequency = <DT_FREQ_K(40)>;
```
yes, 
for example on the stm32wb55rg the typical value is not so far from the min/max values given by the [datasheet](https://www.st.com/resource/en/datasheet/stm32wb55rg.pdf). So using 32000 is enough for the I-watchdog

FRASTM commented:
changing the approach with the PR #44682

So this one might be useless
In this case is the https://github.com/zephyrproject-rtos/zephyr/pull/44182 required

FRASTM commented:
Closing : now replaced by   https://github.com/zephyrproject-rtos/zephyr/pull/44682, https://github.com/zephyrproject-rtos/zephyr/pull/44182, https://github.com/zephyrproject-rtos/zephyr/pull/44683
